uaclient/entitlements/fips.py

```diff
@@ -167,6 +170,46 @@ def install_packages(
         :param cleanup_on_failure: Cleanup apt files if apt install fails.
         :param verbose: If true, print messages to stdout
         """
+
+        # Prior to installing packages, check if the kernel is being downgraded
+        # and if so verify that the user wants to continue
+        our_full_kernel_str = (
+            system.get_kernel_info().proc_version_signature_version
+        )
+        if our_full_kernel_str is None:
+            LOG.warning("Cannot gather kernel information")
+            return
+        our_m = re.search(
+            r"(?P<kernel_version>\d+\.\d+\.\d+)", our_full_kernel_str
+        )
+        fips_kernel_policy = apt.get_apt_cache_policy_for_package("linux-fips")
+        fips_m = re.search(
+            r"Candidate: (?P<kernel_version>\d+\.\d+\.\d+)", fips_kernel_policy
+        )
+        if fips_m is not None and our_m is not None:
+            our_kernel_version_str = our_m.group("kernel_version")
+            fips_kernel_version_str = fips_m.group("kernel_version")
+            our_version = pkg_resources.parse_version(our_kernel_version_str)
+            fips_version = pkg_resources.parse_version(fips_kernel_version_str)
+            if fips_version < our_version:
+                event.info(
+                    messages.KERNEL_DOWNGRADE_WARNING.format(
+                        current_version=our_kernel_version_str,
+                        new_version=fips_kernel_version_str,
+                    )
+                )
+                if not util.prompt_for_confirmation(
```

Comment:
Maybe we could move this logic into the `messaging` module, just to be consistent with the other prompts we have for `fips`.

My idea is:

1. Move this logic into a separate function that will only check if a kernel downgrade will happen
2. Call this function on the messaging module and if it returns `True`, then we create the promp as a `pre_enable` step.

But let me know if you disagree or if I am missing something about the implementation

Reply:
I noticed that this was the standard paradigm when I originally drafted it.  However, for the apt-cache to return the correct fips kernel version, `self.setup_apt_config(silent=silent)` has to run so the kernel downgrade logic effectively needs to fit in the middle of `RepoEntitlement._perform_enable`.  Open to suggestions as to whether this should use a larger refactor or if we want to keep the downgrade checking logic in `FIPSCommonEntitlement.install_packages`.

Reply:
That's a fair point and I agree that we can't use `pre_enable` for that, but we do have the `pre_install` check inside [install_packages](https://github.com/canonical/ubuntu-pro-client/blob/main/uaclient/entitlements/repo.py#L430). At this point, `setup_apt_config` should have already run.

The messaging method should handle this correctly, because we construct the messages dicts everytime we call it. Maybe this is something we could change in the future, but I think right now it should work.

But please let me know if I am still missing something here

Reply:
I think pre_install should work -- let me try rewriting in that format and make sure it works